# Week 2 - Exercises

## Exercise 1: Ready made data vs Custom made data

### Q1

**Centola's study (custom-made data):**

Pros:
- Strong internal validity. Because Centola designed the experiment with random assignment to different network structures, he could isolate the effect of network topology on behavior spread. This allows him to make causal claims, the network structure caused differences in how behavior spread.
- Complete data. Since the experiment was purpose-built for the research question, the data contains exactly the variables needed. There are no missing fields or unmeasured confounders within the experimental setup.
- Full researcher control. Centola controlled the network structure, the behavior being studied, the timing, and the conditions. This eliminates algorithmic confounding, there is no platform algorithm shaping the results behind the scenes.

Cons:
- Small scale. Custom-made experiments are expensive and time-consuming to set up, limiting sample size to around 1,500 participants. This makes it harder to detect small effects or study heterogeneity across subgroups.
- Weak external validity. The social network was artificial, and participants knew they were in an experiment. This *reactivity* (Salganik, Ch. 2.3.3) means their behavior may not reflect how people act in real social networks. It is unclear whether the findings generalize beyond the lab setting.
- Costly. Recruiting participants, building the platform, and running the experiment requires significant time and resources.

**Nicolaides's study (ready-made data):**

Pros:
- Big. The fitness app dataset contains millions of runners over years of activity, enabling the study of rare events, subgroup differences, and small effects (Salganik, Ch. 2.3.1).
- Always-on. The data was collected continuously and longitudinally, allowing researchers to track changes in running behavior over time without needing to plan data collection around specific events (Salganik, Ch. 2.3.2).
- Nonreactive. Runners were not aware their data would be used for research, so their behavior was not influenced by being observed (Salganik, Ch. 2.3.3).
- Strong external validity. The data reflects real people engaging in real behavior within genuine social networks, making the findings more likely to generalize to real-world settings.
- Low cost of data acquisition. The data already existed, no need to recruit participants or build infrastructure.

Cons:
- Weak internal validity. Without random assignment, the study can only show correlations, not causation. The observed "contagion" effect could be driven by confounders such as shared weather, local events, or similar lifestyles among connected individuals.
- Incomplete. The dataset likely lacks important variables such as runners' motivations, income, health conditions, or offline social interactions, information the app was never designed to capture (Salganik, Ch. 2.3.4).
- Algorithmically confounded. The fitness app may use notifications, challenges, leaderboards, or friend suggestions that artificially influence running behavior. These platform-driven effects are largely invisible to the researchers but could be mistaken for genuine social influence (Salganik, Ch. 2.3.8).
- Nonrepresentative. Users of a fitness tracking app are not a random sample of the general population. They are likely wealthier, healthier, younger, and more tech-savvy, which limits the generalizability of the findings (Salganik, Ch. 2.3.6).
- Dirty. The data may contain GPS errors, inactive accounts, or automated activity that does not reflect real running behavior (Salganik, Ch. 2.3.9).

### Q2

The key differences between the two data sources directly affect what conclusions can be drawn from each study.

Centola's custom-made experiment allows for strong causal claims: one can conclude that network structure causes differences in behavior spread, because all other variables were controlled through random assignment. However, because the experiment took place in an artificial setting with participants who knew they were being observed, we cannot be confident that these results hold in real-world social networks. The small scale also limits the ability to explore whether the effect varies across different types of people or contexts.

Nicolaides's study, by contrast, captures real-world behavior at massive scale, giving it strong ecological validity. But the lack of experimental control means the observed "exercise contagion" could have alternative explanations. For example, two friends might both increase their running not because one influenced the other, but because they live in the same city and experienced the same good weather, or because the app sent both of them a notification about a running challenge (algorithmic confounding). The incompleteness of the data, missing variables like motivation or health, makes it impossible to rule out such confounders. Furthermore, because fitness app users are not representative of the general population, the findings may not apply to people who do not use such apps.

In summary, Centola's study trades generalizability for causal certainty, while Nicolaides's study trades causal certainty for scale and realism. Neither approach alone gives a complete picture, which is why Salganik advocates for hybrid approaches that combine the strengths of custom-made and ready-made data.

## Exercise 2: APIs for Computational Social Science research

**Data description:** The Reddit Data API provides access to posts, comments, user profiles, subreddit metadata, and voting scores across Reddit's communities. This makes it useful for CSS research on public discourse, opinion formation, community dynamics, and sentiment analysis, for example, studying how privacy-related events shift discussion in subreddits like r/privacy or r/degoogle.

**Rate limits (free tier):** Free tier allows 100 requests per minute (QPM) for OAuth-authenticated applications, and 10 requests per minute for unauthenticated access. The QPM limit is averaged over a 10-minute window, so burst traffic is briefly allowed. The free tier is available for non-commercial uses such as personal projects and academic research. One important limitation: you can only paginate through about 1,000 posts per subreddit listing, which caps how much historical data you can pull.

**Link:** https://www.reddit.com/dev/api/

## Exercise 3: Find potential Computational Social Scientists

In [ ]:
import requests
import pandas as pd
import time
from rapidfuzz import fuzz

In [32]:
# load the dedup names from week1
names_df = pd.read_csv("week1_ic2s2_2025_researchers.csv")
names = names_df["name"].tolist()
print(f"Number of researchers: {len(names)}")
print(f"First 5: {names[:5]}")

Number of researchers: 1584
First 5: ['Aakriti Kumar', 'Aaron Clauset', 'Aaron D Nichols', 'Aaron Reeves', 'Aaron Schein']


### Single API request

In [ ]:
# test with a random name
test_name = "Duncan Watts"
url = "https://api.openalex.org/authors"
params = {"search": test_name}
response = requests.get(url, params=params)
data = response.json()

print(f"Status: {response.status_code}")
print(f"Number of results: {len(data['results'])}")

Status: 200
Number of results: 25


In [34]:
first = data['results'][0]
print(f"Keys: {first.keys()}\n")
print(f"ID:            {first['id']}")
print(f"Name:          {first['display_name']}")
print(f"Works API URL: {first['works_api_url']}")
print(f"Works count:   {first['works_count']}")
print(f"H-index:       {first['summary_stats']['h_index']}")
print(f"Institutions:  {first.get('last_known_institutions', [])}")

Keys: dict_keys(['id', 'orcid', 'display_name', 'display_name_alternatives', 'relevance_score', 'works_count', 'cited_by_count', 'summary_stats', 'ids', 'affiliations', 'last_known_institutions', 'topics', 'topic_share', 'x_concepts', 'counts_by_year', 'longest_name', 'parsed_longest_name', 'block_key', 'works_api_url', 'updated_date', 'created_date'])

ID:            https://openalex.org/A5000679279
Name:          Duncan J. Watts
Works API URL: https://api.openalex.org/works?filter=author.id:A5000679279
Works count:   233
H-index:       70
Institutions:  [{'id': 'https://openalex.org/I79576946', 'ror': 'https://ror.org/00b30xv10', 'display_name': 'University of Pennsylvania', 'country_code': 'US', 'type': 'education', 'lineage': ['https://openalex.org/I79576946']}, {'id': 'https://openalex.org/I922845939', 'ror': 'https://ror.org/03zzmyz63', 'display_name': 'Philadelphia University', 'country_code': 'US', 'type': 'education', 'lineage': ['https://openalex.org/I922845939']}]


In [ ]:
# why 25 results?
for i, result in enumerate(data['results']):
    print(f"{i+1}. {result['display_name']} | "
          f"relevance: {result['relevance_score']} | "
          f"works: {result['works_count']} | "
          f"h-index: {result['summary_stats']['h_index']}")

1. Duncan J. Watts | relevance: 46833.027 | works: 233 | h-index: 70
2. Duncan J. Watts | relevance: 8257.978 | works: 143 | h-index: 21
3. Andrew Geddes | relevance: 6771.0215 | works: 218 | h-index: 35
4. Juliet Lodge | relevance: 4108.471 | works: 219 | h-index: 20
5. Simon Usherwood | relevance: 2904.6387 | works: 153 | h-index: 15
6. Duncan Watts | relevance: 1521.5914 | works: 29 | h-index: 4
7. Duncan J. Watts | relevance: 1204.0117 | works: 1 | h-index: 1
8. Joseph W. Duncan | relevance: 637.72833 | works: 35 | h-index: 5
9. Duncan J. Watts | relevance: 559.79926 | works: 2 | h-index: 2
10. Duncan Watts | relevance: 395.91037 | works: 36 | h-index: 1
11. DUNCAN J. WATTS | relevance: 332.93042 | works: 1 | h-index: 1
12. Duncan Watts | relevance: 293.15405 | works: 2 | h-index: 1
13. Duncan Watts | relevance: 292.40146 | works: 2 | h-index: 1
14. Duncan Watts | relevance: 292.40146 | works: 2 | h-index: 1
15. Duncan J. Watts | relevance: 166.71219 | works: 1 | h-index: 1
16. Dun

### Obervations from test request

- The API returns 25 results for a single name, why?
- Multiple results share the same name, these are probably duplicates records in OpenAlex but could also be different people.
- Some results dont match the searched name, likely due to matching on collaborators or metadata.
- The relevance score drops sharply after the top result, which suggest the first result is probably what we should go with.

### Helper func for single author

In [ ]:
def get_author_info(name):
    """Query OpenAlex API for an author name and return info for the top result."""
    url = "https://api.openalex.org/authors"
    params = {"search": name}
    response = requests.get(url, params=params)
    
    if response.status_code != 200:
        return None
    
    data = response.json()
    
    if not data['results']:
        return None
    
    top = data['results'][0]
    
    # extract country_code from last known institution (if available)
    institutions = top.get('last_known_institutions', [])
    country_code = institutions[0]['country_code'] if institutions else None
    
    return {
        'searched_name': name,
        'openalex_id': top['id'],
        'display_name': top['display_name'],
        'works_api_url': top['works_api_url'],
        'h_index': top['summary_stats']['h_index'],
        'works_count': top['works_count'],
        'country_code': country_code,
        'relevance_score': top['relevance_score']
    }

In [37]:
# quick test
result = get_author_info("Duncan Watts")
result

{'searched_name': 'Duncan Watts',
 'openalex_id': 'https://openalex.org/A5000679279',
 'display_name': 'Duncan J. Watts',
 'works_api_url': 'https://api.openalex.org/works?filter=author.id:A5000679279',
 'h_index': 70,
 'works_count': 233,
 'country_code': 'US',
 'relevance_score': 46833.027}

### Loop through all names

In [38]:
results = []
failed = []

for i, name in enumerate(names):
    try:
        info = get_author_info(name)
        if info:
            results.append(info)
        else:
            failed.append(name)
    except Exception as e:
        failed.append(name)
        print(f"Error for '{name}': {e}")
    
    # be polite to the API
    time.sleep(0.1)
    
    # progress update every 100 names
    if (i + 1) % 100 == 0:
        print(f"Processed {i + 1}/{len(names)} names ({len(results)} found, {len(failed)} failed)")

print(f"\nDone! Found: {len(results)}, Failed: {len(failed)}")

Processed 100/1584 names (93 found, 7 failed)
Processed 200/1584 names (190 found, 10 failed)
Processed 300/1584 names (286 found, 14 failed)
Processed 400/1584 names (383 found, 17 failed)
Processed 500/1584 names (479 found, 21 failed)
Processed 600/1584 names (575 found, 25 failed)
Processed 700/1584 names (668 found, 32 failed)
Processed 800/1584 names (760 found, 40 failed)
Processed 900/1584 names (853 found, 47 failed)
Processed 1000/1584 names (928 found, 72 failed)
Processed 1100/1584 names (928 found, 172 failed)
Processed 1200/1584 names (928 found, 272 failed)
Processed 1300/1584 names (928 found, 372 failed)
Processed 1400/1584 names (928 found, 472 failed)
Processed 1500/1584 names (928 found, 572 failed)

Done! Found: 928, Failed: 656


In [40]:
# did we hit ratelimit?
test = requests.get("https://api.openalex.org/authors", params={"search": "Duncan Watts"})
print(f"Status: {test.status_code}")
print(f"Headers: {dict(test.headers)}")

Status: 429
Headers: {'Date': 'Tue, 03 Mar 2026 11:30:45 GMT', 'Content-Type': 'application/json', 'Content-Length': '350', 'Connection': 'keep-alive', 'Retry-After': '44955', 'Access-Control-Allow-Origin': '*', 'X-RateLimit-Cost-Required-USD': '0.001', 'X-RateLimit-Credits-Required': '10', 'X-RateLimit-Limit': '10000', 'X-RateLimit-Limit-USD': '1', 'X-RateLimit-Onetime-Remaining': '0', 'X-RateLimit-Prepaid-Remaining-USD': '0', 'X-RateLimit-Remaining': '0', 'X-RateLimit-Remaining-USD': '0', 'X-RateLimit-Reset': '44955', 'access-control-allow-headers': 'Accept, Accept-Language, Accept-Encoding, Authorization, Content-Type', 'access-control-allow-methods': 'GET, HEAD, POST, OPTIONS', 'access-control-expose-headers': 'Cache-Control, X-RateLimit-Limit, X-RateLimit-Remaining, X-RateLimit-Onetime-Remaining, X-RateLimit-Credits-Used, X-RateLimit-Credits-Required, X-RateLimit-Reset, X-RateLimit-Limit-USD, X-RateLimit-Remaining-USD, X-RateLimit-Prepaid-Remaining-USD, X-RateLimit-Cost-USD, X-Rat

### Retry the failed names with free API key

In [ ]:
API_KEY = "mfbuGMGTVcXte8c81bPeio"

def get_author_info(name):
    url = "https://api.openalex.org/authors"
    params = {"search": name, "api_key": API_KEY}
    response = requests.get(url, params=params)
    
    if response.status_code != 200:
        return None
    
    data = response.json()
    
    if not data['results']:
        return None
    
    top = data['results'][0]
    
    institutions = top.get('last_known_institutions', [])
    country_code = institutions[0]['country_code'] if institutions else None
    
    return {
        'searched_name': name,
        'openalex_id': top['id'],
        'display_name': top['display_name'],
        'works_api_url': top['works_api_url'],
        'h_index': top['summary_stats']['h_index'],
        'works_count': top['works_count'],
        'country_code': country_code,
        'relevance_score': top['relevance_score']
    }

In [42]:
results_retry = []
still_failed = []

for i, name in enumerate(failed):
    try:
        info = get_author_info(name)
        if info:
            results_retry.append(info)
        else:
            still_failed.append(name)
    except Exception as e:
        still_failed.append(name)
        print(f"Error for '{name}': {e}")
    
    time.sleep(0.1)
    
    if (i + 1) % 100 == 0:
        print(f"Retry: {i + 1}/{len(failed)} ({len(results_retry)} found, {len(still_failed)} failed)")

print(f"\nRetry done! Found: {len(results_retry)}, Still failed: {len(still_failed)}")

Retry: 100/656 (49 found, 51 failed)
Retry: 200/656 (146 found, 54 failed)
Retry: 300/656 (240 found, 60 failed)
Retry: 400/656 (333 found, 67 failed)
Retry: 500/656 (430 found, 70 failed)
Retry: 600/656 (527 found, 73 failed)

Retry done! Found: 577, Still failed: 79


In [68]:
all_results = results + results_retry
df = pd.DataFrame(all_results)
print(f"Total researchers found: {len(df)}")
print(f"Names with no results: {len(still_failed)}")
df.head()

Total researchers found: 1505
Names with no results: 79


,searched_name,openalex_id,display_name,works_api_url,h_index,works_count,country_code,relevance_score
0,Aakriti Kumar,https://openalex.org/A5002490736,A. K. Tyagi,https://api.openalex.org/works?filter=author.i...,68,741,IN,4352.1860
1,Aaron Clauset,https://openalex.org/A5014647140,Aaron Clauset,https://api.openalex.org/works?filter=author.i...,47,267,US,32359.1660
2,Aaron D Nichols,https://openalex.org/A5089395967,Aaron Nichols,https://api.openalex.org/works?filter=author.i...,3,10,US,1281.5654
3,Aaron Reeves,https://openalex.org/A5039995516,Aaron Reeves,https://api.openalex.org/works?filter=author.i...,48,254,GB,15843.7620
4,Aaron Schein,https://openalex.org/A5021334672,Joshua Maynez,https://api.openalex.org/works?filter=author.i...,12,52,NaN,6179.6610


In [69]:
if still_failed:
    print(f"Examples of names with no results: {still_failed[:15]}")

Examples of names with no results: ['Abdulaziz Alhumaidy', 'Achyutarama R Ganti', 'Adam (Zhengzi) Zhou', 'Adolfo Fuentes-Jofre', 'Alejandro De la fuente Cuesta', 'Amirhossein Nakhaei', 'Andreas Christian Schneider', 'Ardyn Vivienne Olszko', 'Belén C Saldías Fuentes', 'Bogdan Mursan', 'Carten Allen', 'Chris D.R.O. Starke', 'Clara-Gabriela Clipea', 'Cristian E Candia', 'David Mingfei Liu']


### Inspect data quality

In [80]:
df['name_similarity'] = df.apply(
    lambda row: fuzz.ratio(row['searched_name'].lower(), row['display_name'].lower()),
    axis=1
)

print(f"Mean similarity: {df['name_similarity'].mean():.1f}")
print(f"Below 70% similarity: {(df['name_similarity'] < 70).sum()}")
print(f"\nWorst matches:")
df.nsmallest(10, 'name_similarity')[['searched_name', 'display_name', 'name_similarity', 'relevance_score']]

Mean similarity: 89.6
Below 70% similarity: 208

Worst matches:


,searched_name,display_name,name_similarity,relevance_score
399,Elizaveta Sivak,Елизавета Сивак,6.666667,2406.3840
388,Elena Chechik,Елена Чечик,8.333333,459.1528
1213,Saule Bekova,Сауле Бекова,8.333333,2501.7970
1299,Tarak Shah,Eugene Wu,10.526316,3030.9830
1434,Yilun Liu,Chao Gao,11.764706,13745.0740
1107,Rachel Chen,Debbie A. Lawlor,14.814815,8702.8880
545,Hayley Blunden,Nour Kteily,16.000000,3349.5195
837,Lusha Zhu,Andrew S. Kayser,16.000000,4721.3670
835,Luke Stephens,Moira Tait,17.391304,4847.8180
560,Hitoshi Mitsuhashi,Joel A. C. Baum,18.181818,14938.4740


In [ ]:
print(f"Missing country codes: {df['country_code'].isna().sum()} / {len(df)}")
print(f"\nCountry code distribution:")
df['country_code'].value_counts().head(10)


Missing country codes: 291 / 1505

Country code distribution:


country_code
US    339
CN    134
DE    111
IT     71
GB     59
SE     46
NL     44
KR     39
FR     35
CH     31
Name: count, dtype: int64

In [ ]:
df = df[df['name_similarity'] >= 70].copy()
print(f"Final dataset: {len(df)}")
df.to_csv("week2_openalex_authors.csv", index=False)
print("Saved to week2_openalex_authors.csv")

Final dataset: 1297
Saved to week2_openalex_authors.csv


In [82]:
print(f"Started with: {len(names)} names")
print(f"No API results: {len(still_failed)}")
print(f"Dropped (low similarity): {len(all_results) - len(df)}")
print(f"Final dataset: {len(df)}")

Started with: 1584 names
No API results: 79
Dropped (low similarity): 208
Final dataset: 1297


### Reflection Questions

**Which challenges did you encounter? How did you address them?**

We ran into a few issues. The main one was that the OpenAlex search endpoint returns up to 25 results per name, and the top result isn't always the right person. For instance, searching "Aaron Schein" gave us "Joshua Maynez", a completely different researcher. To handle this, we used fuzzy string matching to compare the searched name against the returned `display_name`, and dropped everything below 70% similarity. Another issue was rate limiting, after about 1,000 requests the API started returning 429 errors. Turns out the free tier only allows 1,000 search calls per day, so we registered for a free API key and retried the failed names. We also noticed some returned names were in non-Latin scripts (e.g., Cyrillic), giving misleadingly low similarity scores despite being correct. We accepted losing those as a trade-off. Lastly, 79 names had no results at all, likely due to scraping artifacts from Week 1 or researchers without an OpenAlex profile.

**Choose one problem you faced while collecting the data and describe your solution. Why did you choose this approach, and what impact might it have on your data?**

The biggest challenge was name disambiguation. The API often returned someone other than the person we searched for, sometimes a researcher with a similar name, sometimes someone completely unrelated. We first tried using the `display_name_alternatives` field, which contains name variants and transliterations, to improve matching. But we found that OpenAlex sometimes incorrectly merges different researchers into one record, so unrelated names show up in each other's alternatives. This made the approach unreliable, for example, "Aaron Schein" appeared in the alternatives of "Joshua Maynez". We went with a simpler solution instead: fuzzy matching between the searched name and `display_name`, dropping anything below 70%. This is conservative and means we lose some valid matches (especially Cyrillic names), but it avoids keeping clearly wrong entries in the dataset. We think a slightly smaller but more reliable dataset is better for any downstream analysis.